src/main/java/in/projecteka/consentmanager/link/link/Link.java

```diff
@@ -119,4 +140,91 @@
+                .flatMap(confirmationResult -> {
+                    if (confirmationResult.getError() != null) {
+                        logger.error("[Link] Link confirmation resulted in error {}", confirmationResult.getError());
+                        return Mono.error(new ClientError(HttpStatus.BAD_REQUEST, cmErrorRepresentation(confirmationResult.getError())));
+                    }
+                    if (confirmationResult.getPatient() == null) {
+                        logger.error("[Link] Link confirmation should have returned linked care context details or error caused." +
+                                "Gateway requestId {}", confirmationResult.getRequestId());
+                        return Mono.error(ClientError.invalidResponseFromHIP());
+                    }
+                    return linkRepository.insertToLink(hipId, patientId, patientLinkRequest.getLinkRefNumber(), confirmationResult.getPatient())
+                            .thenReturn(PatientLinkResponse.builder()
+                                    .patient(confirmationResult.getPatient())
+                                    .build());
+                });
+    }
+
+    private ErrorRepresentation cmErrorRepresentation(RespError respError) {
+        Error error = Error.builder().code(ErrorMap.hipToCmError(respError.getCode())).message(respError.getMessage()).build();
+        return ErrorRepresentation.builder().error(error).build();
+    }
+
+    private LinkConfirmationRequest toLinkConfirmationRequest(PatientLinkRequest patientLinkRequest, UUID requestId) {
+        return LinkConfirmationRequest.builder()
+                    .requestId(requestId)
+                    .timestamp(Instant.now().toString())
+                    .confirmation(new TokenConfirmation(patientLinkRequest.getLinkRefNumber(), patientLinkRequest.getToken()))
+                    .build();
+    }
+
+    private long getExpectedFlowResponseDuration() {
+        return serviceProperties.getTxnTimeout();
+    }
+
+    public Mono<Void> onConfirmLink(LinkConfirmationResult confirmationResult) {
+        if(confirmationResult.hasResponseId()) {
+            return linkResults.put(confirmationResult.getResp().getRequestId(), serializeConfirmationFromHIP(confirmationResult));
+        } else {
+            logger.error("[Link] Received a confirmation response from Gateway without original request Id mentioned.{}", confirmationResult.getRequestId());
+            return Mono.error(ClientError.unprocessableEntity());
+        }
+    }
+
+    private String serializeConfirmationFromHIP(LinkConfirmationResult confirmationResult) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return objectMapper.writeValueAsString(confirmationResult);
+        } catch (JsonProcessingException e) {
+            logger.error("[Link] Can not serialize response from HIP", e);
+        }
+        return null;
+    }
```

Comment:
Can you use the Serializer.from and don't change the method to add log, I have done it separate PR.

Reply:
will add to the next PR.